src/database/manager.ts

```diff
@@ -0,0 +1,155 @@
+  ConnectionOptions,
+  createConnection,
+  EntitySchema,
+  getConnectionManager,
+} from "typeorm"
+import { LoggerOptions } from "typeorm/logger/LoggerOptions"
+
+// todo: maybe make connection names match DB names
+// to support multiple connections to different databases
+const CONNECTION_NAME = "default"
+
+export interface ConnectionOptionsOverrides extends Omit<ConnectionOptions, "type"> {
+  database?: string
+}
+
+export interface DatabaseManagerProps extends ConnectionOptionsOverrides {
+  printQueries?: boolean | undefined
+
+  // enable x-ray instrumentation
+  // (may be broken)
+  tracing?: boolean | undefined
+
+  // list of all database entities
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  entities?: (string | Function | EntitySchema<any>)[]
+}
+
+/**
+ * Manager of database connections.
+ * Needs some improvement. Consider it a work in progress.
+ */
+export class DatabaseManager {
+  private _connectionManager?: ConnectionManager
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  entities?: (string | Function | EntitySchema<any>)[]
+  database?: string
+  printQueries?: boolean | undefined
+  tracing?: boolean | undefined
+  connectionOpts: ConnectionOptionsOverrides
+
+  /**
+   * @param entities List of all database entities
+   */
+  constructor(props?: DatabaseManagerProps) {
+    const { entities, printQueries, tracing, database, ...connectionOpts } = props || {}
+    if (entities) this.entities = entities
+    this.printQueries = printQueries
+    this.tracing = tracing
+    if (database) this.database = database as string
+    this.connectionOpts = connectionOpts
+  }
+
+  get connectionManager(): ConnectionManager {
+    if (this._connectionManager) return this._connectionManager
+    this._connectionManager = getConnectionManager()
+    return this._connectionManager
+  }
+
+  /**
+   *
+   * @returns Database connection
+   */
+  public async getConnection(opts?: ConnectionOptionsOverrides): Promise<Connection> {
+    let connection: Connection
+
+    if (this.connectionManager.has(CONNECTION_NAME)) {
+      connection = await this.connectionManager.get(CONNECTION_NAME)
+
+      if (!connection.isConnected) {
+        connection = await connection.connect()
+      }
+    } else {
+      const options = this.getConnectionOptions(opts)
+      connection = await createConnection(options)
+    }
+
+    return connection
+  }
+
+  getConnectionOptions(_opts?: ConnectionOptionsOverrides): ConnectionOptions {
+    // eslint-disable-next-line prefer-const
+    let { database, ...opts } = _opts || {}
+
+    // config
+    let connectionOptions: ConnectionOptions
+
+    // logging
+    const logging: LoggerOptions = ["error"]
+    if (process.env.SQL_ECHO || this.printQueries) logging.push("query") // log queries
+
+    database ||= this.database || process.env[DB_NAME_ENV] || ""
+
+    const baseOptions = {
+      ...this.connectionOpts,
+      entities: this.entities || [],
```

Comment:
eh? Blow up if no entities?

Reply:
Maybe we will want to write scripts that just get a raw DB connection without any ORM layer